### ISSUE TITLE
Manual Test Framework

### ISSUE THREAD

**Issue by tomspilman** [mono/MonoGame#3793]
We should write a framework for manually run unit tests to augment the existing automated unit test framework.

These manual tests would be specifically for features like audio and input which are most easily validated by people.  The trick would be to make sure the test is quick to do and hard to get wrong.  Some core ideas:

  - The test app should be built and distributed as part of the automated build process.
  - The tests should execute in a fixed order.
  - You must complete one test to move to the next one.
  - Should take very little time to complete the tests... less than a few minutes.
  - The results should contain very basic machine info.
  - The tests should work on all our platforms... including phones.
  - Once all the tests are completed it should display the results, spit out a text file, and display a button to copy the results to your clipboard for easy pasting into GitHub.

The tests themselves should be crafted so that the user has to do the least work possible.  We should ask the user to do something... verify the result wasn't an accident by repeating it...  store the result of the test and move on to the next one.  We **should not** depend on the user to judge themselves if the test passed or not.

The framework for this should be a stander MonoGame `Game` with some sort of `TestScreen` class that is overloaded for different tests.  The framework would then just find them via reflection, sort them in some consistent order, and run them recording results.

The usage for these would be in validating PRs.  If a PR changes something that is covered in the manual tests we need people to run thru them on all the target platforms before approval and merge.  We would also run them once before we do a release.

Thoughts?

**Comment by cra0zy** [comment]
I think this sounds like a good idea, also we would need to check e.g. if device has touchscreen before giving the device a touchscreen test.

**Comment by tomspilman** [comment]
> we would need to check e.g. if device has touchscreen before giving the device a touchscreen test.

Right... it would have to be smart like that.  Which should be easy as MonoGame should be able to tell you if the feature is supported on the current machine.

**Comment by craftworkgames** [comment]
This sounds like a great idea. I'm going to have a bit of spare time up my sleeve shortly, so I might get started on it.

 - What tests do you think we need to cover initially?
 - How are we going to cover each platform?
 - Who is going to run them? Or, how do we motivate people to run them and report issues?

Perhaps it should be a kind of MonoGame "demo" that ships with the installer. It could serve as a set of tests during development and a demo to show off what MonoGame can do after release.

**Comment by tomspilman** [comment]
> What tests do you think we need to cover initially?

Test that require a person to operate.  Any test that can or had been done in our existing unit tests:

https://github.com/mono/MonoGame/tree/develop/Test

Should not be included in a manual testing framework... it would be redundant and useless.

> How are we going to cover each platform?

The manual testing framework should be cross platform.  So someone will have to manually run it on each platform to verify it works correctly.  Of course tests that cannot run on some platforms...  like a PC without a touch screen...  should be detected and skipped automatically.


> Who is going to run them?

 Ideally people submitting code changes, core developers like myself, others that feel motivated/compelled to help test.  Again since the manual test app will be included in the build artifacts... anyone should be able to download a zip and run them easily.


> MonoGame "demo" that ships with the installer.

It shouldn't be a demo of any kind... it is strictly a testing tool.  We wouldn't really need to ship it with the installer as by that time it is too late... the build shipped and any bugs need to be reported as such.  Testing is already over.

If you want to work on a MonoGame demo app that is totally a different thing from this.

**Comment by craftworkgames** [comment]
Thanks @tomspilman for the very well written set of requirements.

I guess we want to keep the tests fairly minimal (dependency wise) but to make it usable we are going to need at least a few bits of content. 
 - A font to display instructions on what to do in each test.
 - Buttons to click to interact with the tests.
 - Content for audio tests (sound effect + music)

To get started I think we need to rough out a UI design for a single test. For example, for the audio test I guess you'd have a button to click, listen for the sound and confirm you heard it.
Once the test has been passed or failed you click another button to move onto the next test.

Does that sound like what you had in mind?

**Comment by hach-que** [comment]
Do you mind if I prototype something for this?  Having this would have been helpful for testing all of the platforms during #4282.

**Comment by tomspilman** [comment]
> Do you mind if I prototype something for this? 

Sure... as long as it is a stand alone test framework and not a Protobuild feature.

I have some code I wrote for tests we use on PS4/Vita that might have some good stating points.

> Having this would have been helpful for testing all of the platforms

Well... the most this app should do is starting up a Game instance and doing some basic rendering to display test info and results.  It shouldn't be used to duplicate the existing unit tests.

**Comment by hach-que** [comment]
> Sure... as long as it is a stand alone test framework and not a Protobuild feature.

Yeah, of course? o_o?

**Comment by tomspilman** [comment]
I quickly packed up what I had here:

https://app.box.com/s/43936r5d3fd8xh1jxidvujd13xfkfp16

Had to delete a bunch of source and content that was PS4/Vita specific and couldn't share, so it might not build out of the gate.  Still could have some useful ideas as to how to structure a test app like I outlined above.

**Comment by hach-que** [comment]
I mighhhtt hold off until #4282 because the platform names are changing, and this will necessitate creation of a new project definition.

**Comment by cra0zy** [comment]
I don't know if it's useful, but I had created this: https://youtu.be/KuZZQNRnvYk for gamepad testing a while ago and remembered it today while I was setting up SDL bindings.

**Comment by Tzenchor** [comment]
While my other pulls are waiting for review I have been working on a [manual test framework](https://onedrive.live.com/redir?resid=7C0DE47C61522229!6483&authkey=!ANeywIuLxb0y6MU&ithint=file%2czip), the project is for Visual Studio 2015, but includes a compiled version inside `ManualTestFramework\bin`.
I based it on a UI library I built for XNA and ported it for MonoGame (along with the game). Keep in mind that this library was coded in spanish, and includes our unique character "ñ" and accent marks (I would not mind translating it).

The core of the framework is comprised by the following classes:

- `TestStep`: Defines each test case, uses delegates to work
- `TestWindow`: Core implementation of the tests, similar to nUnit TestFixture. Runs each TestStep in order and closes automatically when all steps are completed. All test windows must extend this one. Includes buttons to allow the user to skip steps if needed
- `MasterWindow`: Main window of the application that allows to select which tests to run and export the results using the same format as the automatic tests.

There's one test implemented, and it is more for testing the framework than a proper test, it is `GamePadTestWindow`, but uses the keyboard.

The interface:
![masterwindow](https://cloud.githubusercontent.com/assets/11319966/14596460/e983caf8-0546-11e6-8ca2-283c45b2a550.PNG)
`MasterWindow`: The left panel contains all detected tests and allows to disable anyone, at this time there's no detection of the system capabilities, but the idea is to disable the tests that cannot be run. 
The right panel shows the test results per `TestWindow`.
The bottom buttons can be clicked or use their hotkeys, "Begin" responds to "B", "Export" to "E", and "Exit" to "Escape", the Export button is disabled until the tests are run, clicking it generates an xml, similar to the nUnit one, and uses a copy of the xsl to generate and open a HTML report like the one from the automated test project.
![gamepadtestwindow](https://cloud.githubusercontent.com/assets/11319966/14596469/f9ee8b4e-0546-11e6-9da0-169d1a69f45a.PNG)
`GamePadTestWindow`: Ignore the black and bluish boxes as they are part of the test. The buttons present can only be activated via mouse and they tag the current step as "Success", "Fail" or "Ignore". "Abort" will tag all remaining tests as Ignore

Caveats:
- There's no test filtering based on device capabilities
- UI library is not prepared for touch input, but only requires implementing it in `GestorEntradaUsuario` (UserInputManager).
- UI library is tested on Windows under XNA, WindowsDX and DesktopGL (works okay with SDL, but with OpenTK mouse coordinates are wrong, and can't use a different window size). On Mono platforms it is untested
- `TestStep` instructions must contain newline characters as there is no text wrapping at this time.

**Comment by tomspilman** [comment]
Great work @Tzenchor !

> The core of the framework is comprised by the following

 Seems like a reasonable start.


> There's no test filtering based on device capabilities

I think first case we should allow for manual skipping on tests that don't work on the current device.  Then it would be nice to detect this automatically in the future.


> UI library

I am really hesitant to add a full UI library for manual tests.  I really think there is little need here past a button for most all situations.

At the very least I would want to be sure we remove anything we're not using and keep the library stripped and at a bare minimum for our needs.

Finally we would really want to relicense the library under the MonoGame copyright/license.

The alternative to all this is a good UI library that is PCL and available from NuGet.  Then at least we don't have to include its source, but I do worry about getting broken if someone makes changes to the public library.

>  Keep in mind that this library was coded in spanish, and includes our unique character "ñ" and accent marks (I would not mind translating it).

We will require that.

**Comment by KonajuGames** [comment]
> I am really hesitant to add a full UI library for manual tests. I really think there is little need here past a button for most all situations.

Agreed.  Drawing text for prompts and a rectangle for a button is all we need.

**Comment by craftworkgames** [comment]
> The alternative to all this is a good UI library that is PCL and available from NuGet.

We are working on a UI library in MonoGame.Extended that would be available as a PCL from NuGet. Unfortunately it's still a work in progress any may not be ready for some time.

Great to see someone working on this though. I'd be happy with whatever approach is taken.

**Comment by Tzenchor** [comment]
> I am really hesitant to add a full UI library for manual tests. I really think there is little need here past a button for most all situations.

I also agree with this. In the end I used the library because it automates many things, and without it the `MasterWindow` would have been a pain to do (specially the lists).

> I think first case we should allow for manual skipping on tests that don't work on the current device.

That is the idea of the `MasterWindow`, you manually untag all test that you don't want to run, once the capabilities detection is on, those that can't be tested would be automatically unchecked and disable so the user can't activate them accidentally.

> PCL compatibility

I have looked at converting the UI library to PCL and it is not compatible (for the record, the target is "profile111"), some controls relay on the old Reflection API to work and would need some changes changed (I have not used any of them here) and others would need some rework as they use functions not available. Concretelly there are 52 errors to make it compliant

> Relicensing

That would not be a problem

> We are working on a UI library in MonoGame.Extended that would be available as a PCL from NuGet. Unfortunately it's still a work in progress and may not be ready for some time.

@craftworkgames if anything of this library can be useful there, take it freely

**Comment by Tzenchor** [comment]
For the record, I have converted the UI library to PCL, the only caveat is that I had to disable the language probing for localization, but that is irrelevant for this

**Comment by tomspilman** [comment]
I'm going to jump back in on this one and get this done once and for all.  We badly need some quality control in our input, audio, and windowing code.
